src/OrbitGl/LineGraphTrack.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "LineGraphTrack.h"
+
+#include <GteVector.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+
+#include "Geometry.h"
+#include "TextRenderer.h"
+#include "TimeGraph.h"
+#include "TimeGraphLayout.h"
+
+namespace orbit_gl {
+
+template <size_t Dimension>
+float LineGraphTrack<Dimension>::GetLabelYFromValues(
+    const std::array<double, Dimension>& values) const {
+  float content_height = this->size_[1] - this->layout_->GetTrackTabHeight() -
+                         this->layout_->GetTrackBottomMargin() - this->GetLegendHeight();
+  float base_y = this->pos_[1] - this->size_[1] + this->layout_->GetTrackBottomMargin();
+  double min = this->GetGraphMinValue();
+  double inverse_value_range = this->GetGraphInverseValueRange();
+  std::array<float, Dimension> normalized_values;
+  std::transform(values.begin(), values.end(), normalized_values.begin(),
+                 [min, inverse_value_range](double value) {
+                   return static_cast<float>((value - min) * inverse_value_range);
+                 });
+
+  if (Dimension == 1) return base_y + normalized_values[0] * content_height;
+  return base_y + content_height / 2.0f;
+}
+
+template <size_t Dimension>
+void LineGraphTrack<Dimension>::DrawSeries(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                           float z) {
+  auto entries_affected_range_result =
+      this->series_.GetEntriesAffectedByTimeRange(min_tick, max_tick);
+  if (!entries_affected_range_result.has_value()) return;
+
+  typename MultivariateTimeSeries<Dimension>::Range& entries{entries_affected_range_result.value()};
+
+  double min = this->GetGraphMinValue();
+  double inverse_value_range = this->GetGraphInverseValueRange();
+
+  auto current_iterator = entries.begin;
+  uint64_t current_time = current_iterator->first;
+  std::array<float, Dimension> current_normalized_values;
+  std::transform(current_iterator->second.begin(), current_iterator->second.end(),
+                 current_normalized_values.begin(), [min, inverse_value_range](double value) {
+                   return static_cast<float>((value - min) * inverse_value_range);
+                 });
+
+  while (current_iterator != entries.end) {
+    auto next_iterator = std::next(current_iterator);
+    uint64_t next_time = next_iterator->first;
+    std::array<float, Dimension> next_normalized_values;
+    std::transform(next_iterator->second.begin(), next_iterator->second.end(),
```

Comment:
I've already seen this code 3 times, could it be replaced by a function?

Reply:
Added a static free function `GetNormalizedValues` for this part. Thanks for suggestion!